File: algorithmic-toolbox/W2/05 Fibonacci Number Again/fibonacci_huge.py
# python3
import sys


def get_fibonacci(n):
    if (n < 1):
        return n

    prev = 0
    curr = 1

    for _ in range(n - 1):
        prev, curr = curr, prev + curr

    return curr
    
def get_pisano_period(m):
    prev = 0
    curr = 1
    
    for i in range(m * m):
        prev, curr = curr, (prev + curr) % m
        if prev == 0 and curr == 1:
            return i + 1
            
def get_fibonacci_huge(n, m):
    remainder = n % get_pisano_period(m)
    return get_fibonacci(remainder) % m

if __name__ == '__main__':
    input = sys.stdin.read();
    n, m = map(int, input.split())
    print(get_fibonacci_huge(n, m))

File: algorithmic-toolbox/W2/05 Fibonacci Number Again/FibonacciHuge.java
import java.util.*;

public class FibonacciHuge {
    private static long pisanoPeriod(long mod) {
        long prev = 0;
        long curr = 1;

        long pisanoPeriod = 0;
        for (int i = 0; i < mod * mod; i++) {
            long next = (prev + curr) % mod;
            prev = curr;
            curr = next;
            if (prev == 0 && curr == 1) {
                pisanoPeriod = i + 1;
                break;
            }
        }
        return pisanoPeriod;
    }

    private static long getFibonacciHuge(long number, long mod) {
        long remainder = number % pisanoPeriod(mod);
        if (remainder < 1)
            return remainder;

        long prev = 0;
        long curr = 1;

        for (int i = 1; i < remainder; i++) {
            long next = (prev + curr) % mod;
            prev = curr;
            curr = next;
        }
        return curr % mod;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        long n = scanner.nextLong();
        long m = scanner.nextLong();
        System.out.println(getFibonacciHuge(n, m));
    }
}